core/controllers/story_editor.py

```diff
@@ -0,0 +1,165 @@
+
+        if version_from_payload != story_version:
+            raise base.BaseHandler.InvalidInputException(
+                'Trying to update version %s of story from version %s, '
+                'which is too old. Please reload the page and try again.'
+                % (story_version, version_from_payload))
+
+    @acl_decorators.can_edit_story
+    def get(self, topic_id, story_id):
+        """Populates the data on the individual story page."""
+        if not feconf.ENABLE_NEW_STRUCTURES:
+            raise self.PageNotFoundException()
+
+        _require_valid_id(story_id)
+        _require_valid_id(topic_id)
+
+        story = story_services.get_story_by_id(story_id, strict=False)
+
+        if story is None:
+            raise self.PageNotFoundException(
+                Exception('The story with the given id doesn\'t exist.'))
+
+        self.values.update({
+            'story': story.to_dict()
+        })
+
+        self.render_json(self.values)
+
+    @acl_decorators.can_edit_story
+    def put(self, topic_id, story_id):
+        """Updates properties of the given story."""
+        if not feconf.ENABLE_NEW_STRUCTURES:
+            raise self.PageNotFoundException()
+
+        _require_valid_id(story_id)
+        _require_valid_id(topic_id)
+        story = story_services.get_story_by_id(story_id, strict=False)
+        if story is None:
+            raise self.PageNotFoundException(
+                Exception('The story with the given id doesn\'t exist.'))
+
+        version = self.payload.get('version')
+        self._require_valid_version(version, story.version)
+
+        commit_message = self.payload.get('commit_message')
+        change_dicts = self.payload.get('change_dicts')
+        change_list = [
+            story_domain.StoryChange(change_dict)
+            for change_dict in change_dicts
+        ]
+        try:
+            story_services.update_story(
+                self.user_id, story_id, change_list, commit_message)
+        except utils.ValidationError as e:
+            raise self.InvalidInputException(e)
+
+        story_dict = story_services.get_story_by_id(story_id).to_dict()
+
+        self.values.update({
+            'story': story_dict
+        })
+
+        self.render_json(self.values)
+
+    @acl_decorators.can_delete_story
+    def delete(self, topic_id, story_id):
+        """Handles Delete requests."""
+        if not feconf.ENABLE_NEW_STRUCTURES:
+            raise self.PageNotFoundException()
+
+        _require_valid_id(story_id)
+        _require_valid_id(topic_id)
+        if not story_id:
+            raise self.PageNotFoundException
+
+        topic = topic_services.get_topic_by_id(topic_id, strict=False)
+        if topic is None:
+            raise self.PageNotFoundException(
+                Exception('The topic with the given id doesn\'t exist.'))
+
+        story_services.delete_story(self.user_id, story_id)
+        canonical_story_ids = topic.canonical_story_ids
```

Comment:
Should be topic_services.delete_story(topic, story_id), and that method should call topic.delete_story() while generating a changelist.

Also make sure to check both lists of story_ids, not just the canonical ones. That way, this generalizes in future. Or alternatively, add a prominent note to the docstring saying "NOTE TO DEVELOPERS: At present, this only works for canonical story IDs." The latter course of action may be better.

Reply:
Done, added the docstring.